### PR TITLE
feat: introduce chain-listener [fixes NET-694 NET-673 NET-674 NET-677 NET-685]

### PR DESCRIPTION
**Pull request by justprosh**
## Description
Subscribe to `ComputeCapacityActivated` events and store them on disk (for now). + subscribe to `newHeads`

## Motivation
Preparation for RXD integration.

## Checklist
- [ ] The code follows the project's coding conventions and style guidelines.
- [ ] All tests related to the changes have passed successfully.
- [ ] Documentation has been updated to reflect the changes (if applicable).
- [ ] All new and existing unit tests have passed.
- [ ] I have self-reviewed my code and ensured its quality.
- [ ] I have added/updated necessary comments to aid understanding.

## Reviewer Checklist
- [ ] Code has been reviewed for quality and adherence to [guidelines](https://doc.rust-lang.org/1.0.0/style/README.html).
- [ ] Tests have been reviewed and are sufficient to validate the changes.
- [ ] Documentation has been reviewed and is up to date.
- [ ] Any questions or concerns have been addressed.

### CODE REVIEW

File: crates/chain-listener/src/listener.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        client: &Client,
+    ) -> eyre::Result<Subscription<JsonValue>> {
+        let subs = client
+            .subscribe("eth_subscribe", rpc_params!["newHeads"], "eth_unsubscribe")
+            .await?;
+
+        Ok(subs)
+    }
+
+    pub async fn subscribe_cc_events(&self, client: &Client) -> eyre::Result<Subscription<Log>> {
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:79]:
```suggestion
    pub async fn subscribe_cc_activated(&self, client: &Client) -> eyre::Result<Subscription<Log>> {
```

**Review by folex — APPROVED** [review]

File: crates/log-utils/src/lib.rs [hunk 2]
```diff
@@ -66,6 +66,7 @@ fn default_directives() -> Vec<Directive> {
         "marine_core::module::marine_module=info",
         "runtime::resource=info",
         "connected_client=debug",
+        "listene=debug",
```
Comment (gurinderu) [inline crates/log-utils/src/lib.rs:69]: listener ?

File: crates/chain-data/src/log.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+use crate::error::ChainDataError;
+use crate::log::LogParseError::{MissingToken, MissingTopic};
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Log {
+    // Log arguments
+    pub data: String,
+    // The block number that contains this log
+    pub block_number: String,
```
Comment (gurinderu) [inline crates/chain-data/src/log.rs:16]: is it String

File: crates/chain-listener/src/event/cc_activated.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+///    bytes32 indexed commitmentId,
+///    uint256 endEpoch,
+///    bytes32[] unitIds,
+///);
+/// ```
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CCActivatedData {
+    #[serde(with = "peerid_serializer")]
+    pub peer_id: PeerId,
+    pub commitment_id: Vec<u8>,
```
Comment (gurinderu) [inline crates/chain-listener/src/event/cc_activated.rs:24]: Make a newtype for that

File: crates/chain-listener/src/event/cc_activated.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    pub end_epoch: U256,
+    pub unit_ids: Vec<Vec<u8>>,
```
Comment (gurinderu) [inline crates/chain-listener/src/event/cc_activated.rs:26]: newtype also

File: crates/chain-listener/src/listener.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    pub fn start(self) -> JoinHandle<()> {
+        let result = tokio::task::Builder::new()
+            .name("ChainListener")
+            .spawn(async move {
+                let client = WsClientBuilder::default()
+                    .build(&self.config.ws_endpoint)
+                    .await
+                    .unwrap_or_else(|_| panic!("Could not connect to chain {}", self.config.ws_endpoint));
+
+                let mut _heads = self.subscribe_new_heads(&client).await.expect("Could not subscribe to new heads");
```
Comment (gurinderu) [inline crates/chain-listener/src/listener.rs:37]: this variable is used

File: crates/chain-listener/src/listener.rs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                let mut cc_events = self.subscribe_cc_activated(&client).await.expect("Could not subscribe to cc events");
+
+                loop {
+                    tokio::select! {
+                        _ = _heads.next() => {},
```
Comment (gurinderu) [inline crates/chain-listener/src/listener.rs:42]: Not sure, but you can make a logging here

Comment (justprosh) [reply crates/chain-listener/src/listener.rs:42]: I will do smth meaningful on the next steps of integration with RandomX

**Pull request merged**